src/anyio/abc/_sockets.py

```diff
@@ -31,6 +32,71 @@
 T_Retval = TypeVar("T_Retval")
 
 
+def _validate_socket(
+    sock_or_fd: socket.socket | int,
+    sock_type: socket.SocketKind,
+    addr_family: socket.AddressFamily = socket.AF_UNSPEC,
+    *,
+    require_connected: bool = False,
+    require_bound: bool = False,
+) -> socket.socket:
+    if isinstance(sock_or_fd, int):
+        try:
+            sock = socket.socket(fileno=sock_or_fd)
+        except OSError as exc:
+            if require_connected:
+                raise ValueError("the socket must be connected") from exc
+            elif require_bound:
+                raise ValueError("the socket must be bound to a local address") from exc
+            else:
+                raise
+    elif isinstance(sock_or_fd, socket.socket):
+        sock = sock_or_fd
+    else:
+        raise TypeError(
+            f"expected an int or socket, got {type(sock_or_fd).__qualname__} instead"
+        )
+
+    try:
+        if require_connected:
+            try:
+                sock.getpeername()
+            except OSError:
+                raise ValueError("the socket must be connected") from None
```

Comment:
why raise `from None` instead of the `OSError` as done above?

Reply:
e.g. if the user expects the socket to be connected because of some precondition, but then the creation fails and just the value error is raised. Then, the user would not know why the socket is not connected.

Does the [`sock.getpeername()`](https://docs.python.org/3/library/socket.html#socket.socket.getpeername) method returns always the remote peer name once it has been connected at least once and does not reflect the current state of the connection? Because if so, then the `ValueError` is just fine

Reply:
`sock.getpeername()` returns the remote address only if the socket is currently in a connected state. Otherwise `OSError` is raised. If the socket is not connected, the `OSError` does not give away any additional information. That's why I chose to omit it.

Reply:
ok. thanks for the explanation

Reply:
Hmm – the socket could also be in a closed state, rather than just not connected. That could justify adding the `OSError` as the `__cause__`.

Reply:
I think it does not "hurt" to add the information. Then you have it when you need it and can be ignored when not needed

Reply:
I updated the code now.